### PR TITLE
AI - More conservative use of ResetPathFinding, some polishing

### PR DESCRIPTION
**Pull request by SafwatHalaby**
A followup to #1935, I made the mobs wiser about choosing when to re-calculate their path, this is an optimization PR, with the exception of fixing a minor jumping / swimming bug, no features were changed. I've got an idea in mind for even further recalculation improvements later.

### CODE REVIEW

File: src/Mobs/Monster.cpp [hunk 1]
```diff
@@ -122,10 +123,16 @@ void cMonster::SpawnOn(cClientHandle & a_Client)
 
 bool cMonster::TickPathFinding(cChunk & a_Chunk)
 {
+
 	if (!m_IsFollowingPath)
 	{
 		return false;
 	}
+	if (m_TicksSinceLastPathReset<1000)
```
Comment (madmaxoft) [inline src/Mobs/Monster.cpp]: spaces

File: src/Mobs/Monster.cpp [hunk 2]
```diff
@@ -122,10 +123,16 @@ void cMonster::SpawnOn(cClientHandle & a_Client)
 
 bool cMonster::TickPathFinding(cChunk & a_Chunk)
 {
+
```
Comment (madmaxoft) [inline src/Mobs/Monster.cpp]: remove the empty line

File: src/Mobs/Monster.cpp [hunk 3]
```diff
@@ -133,16 +140,29 @@ bool cMonster::TickPathFinding(cChunk & a_Chunk)
 		return false;
 	}
 
+	double Deviation = (m_FinalDestination - m_PathFinderDestination).Length();  // Distance between where we're going and where we should go.
+	if (Deviation > 0.25)
+	{
+		if (
+			((GetPosition() - m_PathFinderDestination).Length() < 0.25) ||  // If we reached the last path waypoint.
+			m_TicksSinceLastPathReset > 0.15 * pow((m_FinalDestination - GetPosition()).Length(), 2)  // The further we are, the higher the interval for resetting path.
```
Comment (madmaxoft) [inline src/Mobs/Monster.cpp]: Don't use a `pow` function for integral powers.

File: src/Mobs/Monster.h [hunk 4]
```diff
@@ -185,8 +189,8 @@ class cMonster :
 	/** Returns if the ultimate, final destination has been reached */
 	bool ReachedFinalDestination(void) { return ((m_FinalDestination - GetPosition()).SqrLength() < (m_AttackRange * m_AttackRange)); }
 
-	/** Returns if the intermediate waypoint of m_Destination has been reached */
-	bool ReachedDestination(void) { return ((m_Destination - GetPosition()).SqrLength() < 0.25); }
+	/** Returns if the intermediate waypoint of m_NextWayPointPosition has been reached */
+	bool ReachedDestination(void) { return ((m_NextWayPointPosition - GetPosition()).SqrLength() < 0.25); }
```
Comment (madmaxoft) [inline src/Mobs/Monster.h]: Rename the function accordingly

File: src/Mobs/Monster.cpp (hunk 3 above)
Comment (SafwatHalaby) [inline src/Mobs/Monster.cpp]: What should I use instead?

File: src/Mobs/Monster.h (hunk 4 above)
Comment (SafwatHalaby) [inline src/Mobs/Monster.h]: Done.

File: src/Mobs/Monster.cpp (hunk 2 above)
Comment (SafwatHalaby) [inline src/Mobs/Monster.cpp]: Done.

File: src/Mobs/Monster.cpp (hunk 1 above)
Comment (SafwatHalaby) [inline src/Mobs/Monster.cpp]: Done.

**Comment by SafwatHalaby** [comment]
Squashed.

**Comment by SphinxC0re** [comment]
I don't know if this is related but as I played, it felt like  Zombies weren't following me strictly. They seemed to follow long forgotten paths ;)

**Comment by SafwatHalaby** [comment]
That was true before #1935 was merged. which was 14 hours ago. Did you update your server since then?

File: src/Mobs/Monster.cpp [hunk 5]
```diff
@@ -133,16 +139,30 @@ bool cMonster::TickPathFinding(cChunk & a_Chunk)
 		return false;
 	}
 
+	double Deviation = (m_FinalDestination - m_PathFinderDestination).Length();  // Distance between where we're going and where we should go.
+	if (Deviation > 0.25)
+	{
+		double DistanceToTarget = (m_FinalDestination - GetPosition()).Length();
+		if (
+			((GetPosition() - m_PathFinderDestination).Length() < 0.25) ||  // If we reached the last path waypoint.
+			m_TicksSinceLastPathReset > 0.15 * DistanceToTarget * DistanceToTarget  // Interval is proportional to distance squared.
```
Comment (tigerw) [inline src/Mobs/Monster.cpp]: Perhaps you could replace `DistanceToTarget * DistanceToTarget` with `(m_FinalDestination - GetPosition()).SqrLength()' and add parentheses around the comparison?

Comment (SafwatHalaby) [inline src/Mobs/Monster.cpp]: Didn't know that existed :P
Thanks!

Comment (SafwatHalaby) [inline src/Mobs/Monster.cpp]: Done.

**Comment by SafwatHalaby** [comment]
Build failed, gotta try xoft's lua pre-commit hook.

**Comment by SafwatHalaby** [comment]
The pre-commit hook looks awesome :)
This PR looks good to me. Any remarks?

**Comment by madmaxoft** [comment]
I'm no longer capable of following all the changes, I just trust that you've tested them out properly.

**Comment by SafwatHalaby** [comment]
I am a little confused: the latest commit in this PR is 304bc43 while Worktycho merged f71d693. But they seem identical, what's this?

**Comment by SafwatHalaby** [comment]
Nevermind my last commit (deleted)

**Comment by jammet** [comment]
Is this about monsters, or animals, too?

**Comment by worktycho** [comment]
In the MCServer inheritance hierarchy any creature that is not a player is a monster, and this covers monsters, so it should improve animals.

**Comment by SafwatHalaby** [comment]
Yes, all living things which are not players are affected.

**Comment by SafwatHalaby** [comment]
Strictly speaking, zombies aren't among the living but you get my point :wink:

**Pull request merged**